File: 1036-rotting-oranges/rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        # Get the dimensions of the grid
        rows = len(grid)
        cols = len(grid[0])

        # Define possible 4-directional moves (down, up, right, left)
        neighbours = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        # Initialize a queue for BFS, a set to track visited cells, and a counter for oranges
        queue = deque()
        seen = set()
        num_of_oranges = 0

        # Initialize the minimum time to -1 to indicate no time has been taken yet
        mintime = -1

        # Traverse the grid to count oranges and identify initially rotten oranges
        for i in range(rows):
            for j in range(cols):
                # Count all oranges (1 for fresh, 2 for rotten)
                if grid[i][j] > 0:
                    num_of_oranges += 1
                    # Add rotten oranges to the queue with their position and starting time 0
                    if grid[i][j] == 2:
                        queue.append((i, j, 0))

        # If there are no oranges, return 0 as no rotting process is needed
        if num_of_oranges == 0:
            return 0

        # BFS to rot the oranges
        while queue:
            xpos, ypos, time = queue.popleft()

            # Skip if the cell has already been visited
            if (xpos, ypos) in seen:
                continue
            
            # Mark the cell as visited
            seen.add((xpos, ypos))

            # Check if all oranges have been reached, and update the minimum time
            if len(seen) == num_of_oranges and ((mintime > 0 and time < mintime) or mintime == -1):
                mintime = time

            # Explore each neighbor to spread the rot to adjacent fresh oranges
            for n in neighbours:
                new_x, new_y = xpos + n[0], ypos + n[1]
                # Ensure the neighbor is within grid bounds, is a fresh orange, and hasn’t been visited
                if 0 <= new_x < rows and 0 <= new_y < cols and grid[new_x][new_y] == 1 and (new_x, new_y) not in seen:
                    # Append the neighbor to the queue with incremented time
                    queue.append((new_x, new_y, time + 1))

        # Return the minimum time if all oranges were reached, otherwise return -1 (if some remain fresh)
        return mintime

File: 1036-rotting-oranges/rotting-oranges.java
class Solution {

    public int orangesRotting(int[][] grid) {
        
        ArrayDeque<int[]> queue = new ArrayDeque<>();
        int rows = grid.length;
        int cols = grid[0].length;
        int[][] neighbours = {{0,1},{0,-1},{1,0},{-1,0}};
        int count = 0;
        

        for (int row = 0; row < rows; row++){
            for (int col = 0; col < cols; col++){

                if (grid[row][col] == 2){
                    int[] orange = new int[2];
                    orange[0] = row;
                    orange[1] = col;
                    queue.addLast(orange);
                } else if (grid[row][col] == 1) count += 1;
            }
        }

        if (count == 0) return 0;

        
        int time = -1;


        while (!queue.isEmpty()){
            int size = queue.size();
            

            for (int idx = 0 ;idx < size; idx ++){
                 int[] current = queue.pollFirst();

                 int r = current[0];
                 int c = current[1];

                 for (int[] n : neighbours){
                    int nr = r + n[0];
                    int nc = c + n[1];

                    if (nr < 0 || nr >= rows || nc < 0 || nc >= cols || grid[nr][nc] != 1 ) continue;

                    grid[nr][nc] = 2;
                    count -= 1;
                    queue.addLast(new int[]{nr,nc});
                 }
            }

            time += 1;
        }


        return (count == 0 ? time : -1);
    }
}